// parcel-bundler/lib/packagers/HTMLPackager.js
'use strict';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const Packager = require('./Packager');
const posthtml = require('posthtml');
const path = require('path');
const urlJoin = require('../utils/urlJoin');

class HTMLPackager extends Packager {
  addAsset(asset) {
    var _this = this;

    return _asyncToGenerator(function* () {
      let html = asset.generated.html || '';

      // Find child bundles that have JS or CSS sibling bundles,
      // add them to the head so they are loaded immediately.
      let siblingBundles = Array.from(_this.bundle.childBundles).reduce(function (p, b) {
        return p.concat([...b.siblingBundles.values()]);
      }, []).filter(function (b) {
        return b.type === 'css' || b.type === 'js';
      });

      if (siblingBundles.length > 0) {
        html = posthtml(_this.insertSiblingBundles.bind(_this, siblingBundles)).process(html, { sync: true }).html;
      }

      yield _this.dest.write(html);
    })();
  }

  getHeadContent(tree) {
    let head = find(tree, 'head');
    if (!head) {
      let html = find(tree, 'html');
      head = { tag: 'head' };
      html.content.unshift(head);
    }

    if (!head.content) {
      head.content = [];
    }

    return head;
  }

  insertSiblingBundles(siblingBundles, tree) {
    let head = this.getHeadContent(tree);

    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = siblingBundles[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        let bundle = _step.value;

        if (bundle.type === 'css') {
          head.content.push({
            tag: 'link',
            attrs: {
              rel: 'stylesheet',
              href: urlJoin(this.options.publicURL, path.basename(bundle.name))
            }
          });
        } else if (bundle.type === 'js') {
          head.content.push({
            tag: 'script',
            attrs: {
              src: urlJoin(this.options.publicURL, path.basename(bundle.name))
            }
          });
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }
}

function find(tree, tag) {
  let res;
  tree.match({ tag }, node => {
    res = node;
    return node;
  });

  return res;
}

module.exports = HTMLPackager;
